src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java

```diff
@@ -738,6 +741,21 @@ static String getPropertyOnlyName(String name, Logger logger) {
         return result;
     }
 
+    static synchronized void addConnRef(UUID clientConnectionId) {
+        if (null != connRef && !connRef.contains(clientConnectionId)) {
+            connRef.add(clientConnectionId);
+        }
+    }
+
+    static synchronized void removeConnRef(UUID clientConnectionId) {
+        if (null != connRef && connRef.contains(clientConnectionId)) {
+            connRef.remove(clientConnectionId);
+            if (0 == connRef.size()) {
```

Comment:
remove() actually returns true or false depending on whether an element was removed. An interesting improvement might be to check whether the remove actually did remove an element via `if (connRef.remove(clientConnectionId && 0 == connRef.size())`, which would fail fast in the event that an element wasn't removed. Maybe it creates potential for errors (if for some reason in the future we don't only remove references here, but only look to kill thread here) and the current implementation is performant enough that it doesn't matter. Not necessarily a needed change, but could be considered.